### PR TITLE
Add GPU assisted validation of device memory in VkCmdDrawIndirect calls

### PR DESCRIPTION
**Pull request by Tony-LunarG**
This replaces #2744 and in addition to VUIDs 03121 and 03122 adds the following:
- The layer no longer replaces the validated draw's count buffer and validation does not try to abort the draw by writing zero into it.
- Adds checks and tests for indexed indirect draws (VUIDs 03153 / 03154)
- Adds script to compile pre-draw validation shader and share constants between layer and shader
- Adds checks and test for maxDrawIndirectCount limit (VUID 02717 x2)
- Adds checks and tests for firstInstance values (VUIDs 00478 / 00530)

### CODE REVIEW

File: layers/gpu_validation.cpp [hunk 1]
```diff
@@ -1246,11 +1281,16 @@ void GpuAssisted::PreCallRecordCreateShaderModule(VkDevice device, const VkShade
         csm_state->instrumented_create_info.codeSize = csm_state->instrumented_pgm.size() * sizeof(unsigned int);
     }
 }
-
+#include "gpu_pre_draw_constants.h"
```
Comment (ncesario-lunarg) [inline layers/gpu_validation.cpp:1284]: May be precedence for this, but these seem a little out of place in the middle of the file (as opposed to the beginning where the rest of the `#include` directives are).

File: layers/generated/gpu_pre_draw_shader.h [hunk 2]
```diff
@@ -0,0 +1,483 @@
+#include <stdint.h>
```
Comment (ncesario-lunarg) [inline layers/generated/gpu_pre_draw_shader.h:1]: Should probably have some sort of a guard (e.g., `#pragma once`/`#ifdef`).

File: layers/gpu_pre_draw_constants.h [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// * Author: Tony Barbour <tony@lunarg.com>
+//
+
+#define _kInstErrorPreDrawValidate 8
```
Comment (ncesario-lunarg) [inline layers/gpu_pre_draw_constants.h:20]: Also should probably have a header guard.

**Review by ncesario-lunarg — COMMENTED** [review]
Couple comments/suggestions, but looks good!

This approach seems better, not modifying the original draw. Though, I wonder if by hitting this validation case an app could crash in the driver, and then they never end up seeing the validation message?

**Comment by Tony-LunarG** [comment]
> Couple comments/suggestions, but looks good!
> 
> This approach seems better, not modifying the original draw. Though, I wonder if by hitting this validation case an app could crash in the driver, and then they never end up seeing the validation message?

Yes, it is true that an invalid draw that isn't aborted by validation can cause a DEVICE_LOST, which would cause the validation error to go unreported.  I've noted that in the GPU-AV projects list for consideration for future work

File: layers/generated/gpu_pre_draw_shader.h (hunk 2 above)
Comment (jeremyg-lunarg) [reply layers/generated/gpu_pre_draw_shader.h:1]: and copyright banner (other generated files do).

**Comment by jeremyg-lunarg** [comment]
So this version gets rid of all the PipelineBarrier madness? nice!

File: scripts/generate_spirv.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    return re.sub("^[^a-zA-Z_]+", "_", s)
+
+def compile(filename, tmpfile):
+    # invoke glslangValidator
+    try:
+        args = [executable, "-V", "-H", "-o", tmpfile, filename]
+        output = subprocess.check_output(args, universal_newlines=True)
+    except subprocess.CalledProcessError as e:
+        print(e.output, file=sys.stderr)
+        exit(1)
```
[the remark this answers is not in the document]
Comment (Tony-LunarG) [reply scripts/generate_spirv.py]: Done

File: scripts/generate_spirv.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+import subprocess
+import struct
+import re
+
+SPIRV_MAGIC = 0x07230203
+COLUMNS = 4
+INDENT = 4
+
+in_filename = sys.argv[1]
+out_filename = sys.argv[2] if len(sys.argv) > 2 else None
```
[the remark this answers is not in the document]
Comment (Tony-LunarG) [reply scripts/generate_spirv.py]: Fixed

File: layers/gpu_validation.cpp (hunk 1 above)
Comment (Tony-LunarG) [reply layers/gpu_validation.cpp]: Moved

File: layers/gpu_pre_draw_constants.h (hunk 3 above)
Comment (Tony-LunarG) [reply layers/gpu_pre_draw_constants.h:23]: Fixed

**Review by jeremyg-lunarg — APPROVED** [review]

**Review by ncesario-lunarg — APPROVED** [review]

**Pull request merged**